src/components/matrix/virtualized-grid.tsx

```diff
@@ -68,73 +123,344 @@ const VirtualizedGrid = (
+  if(!yTree){
+    if(rowListWidth < rowHeaderWidth){
+      rowListWidth = rowHeaderWidth;
+      rowHeaderScrollable = false;
+    }
+  }
+  // Calculate the max height for the column headers when it is scrollable and non-tree component
+  let columnListHeight = colHeaderScrollableMaxHeight===-1 ? xDataMaxLength * columnFontCharWidth : colHeaderScrollableMaxHeight;
+  if(!xTree){
+    if(columnListHeight < columnHeaderHeight){
+      columnListHeight = columnHeaderHeight;
+      colHeaderScrollable = false;
+    }
+  }
+
+  /**
+   * Calculate the y position of the horizontal tree view
+   * In particularly, Get max layer and title to help estimate the position
+   */
+  function getMaxLayersAndTitleLength(treeNodes: TreeNode[]) {
+    let maxLayers = 0;
+    let maxTitleLength = 0;
+
+    const traverseTree = (node: TreeNode, depth: number) => {
+      maxLayers = Math.max(maxLayers, depth);
+      maxTitleLength = Math.max(maxTitleLength, node.title.length);
+
+      if (node.children && node.children.length > 0) {
+        node.children.forEach((child) => {
+          traverseTree(child, depth + 1);
+        });
+      }
+    };
+
+    treeNodes.forEach((node) => {
+      traverseTree(node, 1);
+    });
+
+    return { maxLayers, maxTitleLength };
+  }
+
+  /**
+   * Initialize the dummy scrollbar at the bottom of the row headers
+   */
+  useEffect(() => {
+    if (rowHeaderScrollable && gridContainerRef.current && rowScrollBarRef.current) {
+      addBottomHorizontalScroll(gridContainerRef.current, rowScrollBarRef.current);
+    }
+  }, []);
+
+  /**
+   * Initialize the dummy scrollbar at the right of the column headers
+   */
+  useEffect(() => {
+    if (colHeaderScrollable && gridContainerRef.current && columnScrollBarRef.current) {
+      addRightVerticalScroll(gridContainerRef.current, columnScrollBarRef.current);
+    }
+  }, []);
+
+  /**
+   * Initialize the inner vertical position of the horizontal tree
+   */
+  useEffect(() => {
+    if (xTreeNodes) {
+      const { maxLayers, maxTitleLength } = getMaxLayersAndTitleLength(xTreeNodes);
+      // Font size is 14px in this page
+      const yPositionXTree = maxLayers * maxTitleLength * 14;
+      setScrollTreeYIniPos(yPositionXTree);
+      setScrollTreeY(yPositionXTree);
+      // mark this is the initial page loading
+      setIsRefresh(true);
+    }
+  }, [xTreeNodes]);
+
+  /**
+   * Filter the visiable rows and columns in GridData whenever visiableRowNodes or visiableColNodes changes
+   * which means there are tree nodes expand or collapse
+   */
+  useEffect(() => {
+    // If there is no tree data, then no need to filter visiable rows and columns because they all are visiable
+    if (!xTree && !yTree) return;
+
+    let newData: ParsedGridCell[][] = data;
+    const lastRow: ParsedGridCell[] = data[data.length - 1]; // Store a copy of the last row
+    if (xTree && yTree) {
```

Comment:
I don't think you had a chance to go over this. What you're doing in each block looks similar and we most probably can just refactor it. You should also add inline comments and explain what each one is doing.